✨ [bento][amp-accordion] Add display-locking feature to bento accordion

krdwan opened the pull request:
Add new display locking feature to [bento][accordion].  This supports `ctrl + f` finding of content that is in a collapsed accordion section.  If the content is found, the accordion section will expand.

This only works for new Chromium Canary browser.  To check if the browser supports this, type this in the console: `CSS.supports("content-visibility","hidden-matchable")` (must return true)

API -
AMP - The experiment `amp-accordion-display-locking` must be set to `true`
Preact - The top-level Accordion component must have prop `experimentDisplayLocking` set to `true`
(additionally in both cases, the above CSS property must be supported by the browser)

Implemented by including the following CSS properties on a collapsed content section.
```
content-visibility: hidden-matchable !important;
display: block !important;
```

11/18 - Storybooks added
11/19 - Removed `hidden` attribute
11/19 - Tests in progress

krdwan commented on extensions/amp-accordion/1.0/accordion.js:
```diff
@@ -61,6 +61,7 @@ function AccordionWithRef(
     as: Comp = 'section',
     expandSingleSection = false,
     animate = false,
+    experimentDisplayLocking = false,
```
Top level Preact prop, wanted the name to be closely matched with amp experiment name which is currently set to `amp-accordion-display-locking` from `0.1`

krdwan commented on extensions/amp-accordion/1.0/accordion.js:
```diff
@@ -397,16 +415,47 @@ export function AccordionContent({
+    return () => element.removeEventListener('beforematch', beforeMatchHandler);
+  }, [expandHandler, experimentDisplayLocking]);
+
   useLayoutEffect(() => {
```
What does this block do?

1. Checks that the current environment supports the CSS property `'content-visibility' : 'hidden-matchable'`.  If the browser does not support this, we must exit and not support the feature.  Then we set a state variable within the component to track.

2. If the CSS property is supported AND top-level prop is true, then we add an eventListener to listen for `ctrl + f` finds.  The event listener will cause the content section to expand if it hears anything (AND the component is currently collapsed).

krdwan commented on extensions/amp-accordion/1.0/accordion.js:
```diff
@@ -426,7 +475,9 @@ export function AccordionContent({
+      className={`${className} ${classes.sectionChild} ${classes.content} ${
+        !expanded && useDisplayLocking ? classes.contentHiddenMatchable : ''
+      }`}
       id={contentId}
```
If the display locking feature is enabled and the browser supports the necessary CSS property.  Then add the CSS classes needed for display locking feature.

krdwan commented on extensions/amp-accordion/1.0/amp-accordion.css:
```diff
@@ -29,3 +29,12 @@
+  * {
+  content-visibility: hidden-matchable !important;
+  display: block !important;
+}
```
The CSS for supporting display locking in AMP mode is separate from Preact CSS (handled by JSS file).  This follows our current pattern of handling AMP CSS separately allowing for the publisher to easily overwrite to customize.  Although in this case, we probably do not want to have them overwrite, I think it still makes sense to follow the existing pattern here.

dvoytenko replied:
Another point of duplication here. Ok for now, but let's leave a comment explaining the duplication so that we can revisit it later.

krdwan replied:
Added a comment in the latest revision.

caroqliu commented on extensions/amp-accordion/1.0/accordion.js:
```diff
@@ -397,16 +415,47 @@ export function AccordionContent({
+      return;
+    }
+    const element = ref.current;
+    const win = element.ownerDocument.defaultView;
```
Does this need `toWin`?

krdwan replied:
Hmm, I think it does not in this case.  My understanding based on the comments of `toWin` is that it is primarily used for casting `Window|undefined` to `Window` since many amp internal functions require the `Window` type.  Whereas in this case, we are not calling any of these amp internal functions.  Let me know if I am missing something though!

dvoytenko replied:
This should be either `toWin` or `if (!win) return`.

caroqliu replied:
The issue is if it is `undefined`, calling `undefined.CSS.supports` will throw.

krdwan replied:
Updated to `if (!win) return`.

caroqliu commented on extensions/amp-accordion/1.0/accordion.js:
```diff
@@ -397,16 +415,47 @@ export function AccordionContent({
+      'hidden-matchable'
+    );
+    setContentVisibility(supportsContentVisibility);
+    if (!experimentDisplayLocking || !supportsContentVisibility) {
```
We can move the `!experimentDisplayLocking` guard with the earliest `return` condition.

krdwan replied:
I agree the component will do the same thing in the end.  One benefit of having it where it is though is that the `contentVisibility` state variable will be set to the correct value.  Ultimately, we need both the `contentVisibility` and `experimentDisplayLocking` to be true for the component to turn on the display locking behavior so regardless the final behavior will be the same, but I do like keeping the `contentVisibility` state variable in sync with the environment.  Thoughts?

caroqliu replied:
Calling `setContentVisibility` without reading the `contentVisibility` state will cause an unnecessary render. Not a big deal but something to be aware of. If we flip `experimentDisplayLocking` then this effect will run again and ensure the proper value is set at that point. I don't feel strongly though.

krdwan replied:
Sounds good.

caroqliu commented on extensions/amp-accordion/1.0/accordion.js:
```diff
@@ -397,16 +415,47 @@ export function AccordionContent({
+      return;
+    }
+    const beforeMatchHandler = () => {
+      if (element.hasAttribute('hidden')) {
```
Why is `hidden` important to check here? Why don't we want to expand the section in all cases, where an already expanded content element is just noop? Can we i.e. reuse `expand` or give `expandHandler` an `opt_forceExpand`?

Also, I assumed the `content-visibility: hidden-matchable` content is inside a `[hidden]` element, it shouldn't be findable. I manually checked this and it looks like the `display: block !important` overrides this. Interesting.

krdwan replied:
Regarding checking the `hidden`, the `expandHandler` is actually a toggler.  And it's a toggler because it is re-used from the `header`'s onClick so it has to handle both expand and collapse action.

Yes, the we could give it some type of `opt-forceExpand`, although in this case, I think it's easier to check whether the `hidden` attribute is there or not.  The primary reason I prefer this is because it keeps all of the logic for handling the expansion based on display locking in one place, instead of moving part of it up to the definition of the `expandHandler` (which is in the Section).  And also the `hidden` attribute is a reliable indicator.  Another reason is that the `expandHandler` does not have any dependencies on the current `expanded` state of the current section.  So adding in the `opt-forceExpand` into this function would require the function to also track the `expanded` state, which increases complexity of this function.  In my opinion, it's cleaner to check for `hidden` then to update the logic of `expandHandler` with potentially messier code.

Regarding the `display: block !important` very true, very interesting!

dvoytenko replied:
With display locking we can no longer use the `hidden` attribute.

krdwan replied:
Removed `hidden` attribute in latest update.  Checking for `expanded` attribute on parent in this part now.

caroqliu commented on extensions/amp-accordion/1.0/amp-accordion.js:
```diff
@@ -47,6 +47,10 @@ const CONTENT_SHIM_PROP = '__AMP_C_SHIM';
 
+const isDisplayLockingEnabledForAccordion = (win) =>
+  isExperimentOn(win, 'amp-accordion-display-locking') &&
+  win.document.body.onbeforematch !== undefined;
```
It shouldn't be necessary to check `onbeforematch` here since the Preact layer checks for `content-visibility: hidden-matchable`

krdwan replied:
I agree it's not necessary.  I just figured I'd include it to match with `0.1`.  One benefit of having though is that in the case that the experiment is on, but the `onbeforematch` is `undefined`, we won't be passing in the prop `experimentDisplayLocking` as `true` to the bento component.  Agreed that the bento component still does the right thing since it will still check for `content-visibility: hidden-matchable`, but I figured it wouldn't hurt to check on amp side as well.

Open to updating this one though, what do you think?

dvoytenko replied:
I'm afraid we have to keep it for now b/c we're duplicating CSS at this time. Once we remove CSS duplication, this can go as well, but by then we're likely to have launched the experiment anyway.

krdwan replied:
Sounds good

caroqliu reviewed (commented):
Left a couple comments/questions. Great job with this change. :) I look forward to trying it out in the Storybook!

Btw you can also try this out in the Chrome Stable browser by t urning on `Experimental Web Platform features` in chrome://flags/

krdwan commented on extensions/amp-accordion/1.0/storybook/DisplayLocking.amp.js:
```diff
@@ -0,0 +1,72 @@
+  parameters: {
+    extensions: [{name: 'amp-accordion', version: '1.0'}],
+    experiments: ['amp-accordion-bento', 'amp-accordion-display-locking'],
+  },
```
Added new storybook with the required `amp-accordion-display-locking` experiment enabled.

dvoytenko replied:
This is a good idea.

dvoytenko commented on extensions/amp-accordion/1.0/accordion.js:
```diff
@@ -397,16 +415,47 @@ export function AccordionContent({
+  } = useContext(SectionContext);
   const classes = useStyles();
+  const [contentVisibility, setContentVisibility] = useState(false);
+  const useDisplayLocking = contentVisibility && experimentDisplayLocking;
```
Can't call vars with `use` prefix - those are reserved for hooks only in React.

krdwan replied:
Updated to `enableDisplayLocking`

dvoytenko commented on extensions/amp-accordion/1.0/accordion.js:
```diff
@@ -397,16 +415,47 @@ export function AccordionContent({
+    }
+    const element = ref.current;
+    const win = element.ownerDocument.defaultView;
+    const supportsContentVisibility = win.CSS.supports(
```
I thought we had some issues with checking `CSS.supports` for this due to a Chrome bug? /cc @caroqliu

caroqliu replied:
Appears to be fixed: https://bugs.chromium.org/p/chromium/issues/detail?id=1134237#c11

dvoytenko commented on extensions/amp-accordion/1.0/accordion.jss.js:
```diff
@@ -39,10 +39,16 @@ const header = {
 const content = {};
 
+const contentHiddenMatchable = {
+  'content-visibility': 'hidden-matchable !important',
```
Let's use `contentVisibility` in this format.

krdwan replied:
Updated in latest revision.

dvoytenko commented on extensions/amp-accordion/1.0/amp-accordion.css:
```diff
@@ -29,3 +29,12 @@
   border: solid 1px #dfdfdf;
 }
+
+amp-accordion.i-amphtml-display-locking > section:not([expanded]) > :last-child,
```
We will need some code here for `hidden` as well, since we will no longer be able to use the `hidden` attribute.

krdwan replied:
Added some additional CSS after removing `hidden` in latest revision.

krdwan commented:
Removed `hidden` attribute, and replaced with needed CSS in latest revision.

dvoytenko commented on extensions/amp-accordion/1.0/accordion.js:
```diff
@@ -397,16 +415,53 @@ export function AccordionContent({
+    experimentDisplayLocking,
+  } = useContext(SectionContext);
   const classes = useStyles();
+  const [contentVisibility, setContentVisibility] = useState(false);
```
Let's rename to `supportsContentVisibility`.

krdwan replied:
Updated in latest.

dvoytenko commented on extensions/amp-accordion/1.0/accordion.js:
```diff
@@ -397,16 +415,53 @@ export function AccordionContent({
+    }
+
+    const beforeMatchHandler = () => {
+      if (!element.parentElement.hasAttribute('expanded')) {
```
Seems weird to dig into the parent element here. Can we instead derive the logic from our own state? Or this element?

krdwan replied:
I used `aria-hidden` in the latest revision.  Using classlist required a dependency on the class names generated by JSX, so opted against that, but open to suggestion.

dvoytenko commented on extensions/amp-accordion/1.0/accordion.jss.js:
```diff
@@ -39,10 +39,21 @@ const header = {
+
+const contentHiddenMatchable = {
+  contentVisibility: 'hidden-matchable !important',
+  display: 'block !important',
```
How do we know that `contentHiddenMatchable {display: block !important}` will override the one above `contentHidden {display: none !important}`? I think this is fragile. I think you should either use only one of these classes when you compute the `className`, or you can do a compound selector here, which would guarantee the correct CSS specificity.

krdwan replied:
Updated logic to use only one of these classes.

dvoytenko commented on extensions/amp-accordion/1.0/amp-accordion.css:
```diff
@@ -29,3 +29,20 @@
+  > section:not([expanded])
+  > :last-child
+  * {
+  content-visibility: hidden-matchable !important;
```
Does this not kill the animation?

dvoytenko replied:
Confirmed. Animations are broken:

- In Preact collapsing animation is broken
- In AMP expanding animation is broken

krdwan replied:
Fixed animations in the latest revision: https://github.com/ampproject/amphtml/pull/31227/commits/aa5a335db4af265be3c1011c10909307d4789818

The strategy is to add a class `i-amphtml-animating` to the content element while the animations are occuring.  CSS for hidden content will ignore this class `(:not(.i-amphtml-animating))`.

dvoytenko commented on extensions/amp-accordion/1.0/amp-accordion.css:
```diff
@@ -29,3 +29,20 @@
+
+amp-accordion > section:not([expanded]) > :last-child,
+amp-accordion > section:not([expanded]) > :last-child * {
+  display: none !important;
```
Same: it'd seem like it'd break the animation.

krdwan commented on extensions/amp-accordion/1.0/animations.js:
```diff
@@ -26,51 +26,58 @@ const COLLAPSE_CURVE = 'cubic-bezier(0.39, 0.575, 0.565, 1)';
+        }
+      );
+    },
+    () => {
```
The changes are to add the callback below, and add the class on line 45.  Adding the callback caused the whole function to get indented.

krdwan commented:
Latest changes:

1. Change `expandHandler` to `toggleHandler` in `accordion.js` and add an optional argument that forces `expand` or `collapse` only.
2. Made default css of a section to be `display:block` (w/o `important` tag).  This way it can be easily overwritten when it is hidden.
3. Updated storybook with `div` for all `content` elements to be consistent.
4. Updated unit tests.

https://github.com/ampproject/amphtml/pull/31227/commits/6f6b04802d9e27e7e4bad25d0cba9abe8b56c5d0

dvoytenko commented on extensions/amp-accordion/1.0/accordion.js:
```diff
@@ -109,11 +110,12 @@ function AccordionWithRef(
-    (id) => {
+    (id, opt_expand) => {
       setExpandedMap((expandedMap) => {
+        const newExpanded = opt_expand || !expandedMap[id];
```
It's best to keep the structure to the classical toggle, which is either the `opt_expand` is `undefined` and we compute it, or it's `true|false` and then we use the value as is. In simple terms this becomes:

```
const newExpanded = opt_expand ?? !expandedMap[id];
```

I.e. `??` instead of `||`.

krdwan replied:
Thanks, yes, changed to `??` to have logic as classic toggle.  Thanks!

dvoytenko commented on extensions/amp-accordion/1.0/accordion.js:
```diff
@@ -299,22 +310,25 @@ export function AccordionSection({
+        toggleExpanded(id, opt_expand);
+      } else {
+        setExpandedState((prev) => {
+          const newValue = opt_expand || !prev;
```
Same with `??`

krdwan replied:
Updated in latest revision.

dvoytenko commented on extensions/amp-accordion/1.0/accordion.js:
```diff
@@ -397,16 +419,57 @@ export function AccordionContent({
+      'hidden-matchable'
+    );
+    setSupportsContentVisibility(newSupportsContentVisibility);
+    if (!experimentDisplayLocking || !newSupportsContentVisibility) {
```
Perhaps we should test up top `if (experimentDisplayLocking)` and shortcircuit out if not and avoid the `CSS.supports` computation?

krdwan replied:
Updated in the latest revision.

dvoytenko commented on extensions/amp-accordion/1.0/amp-accordion.css:
```diff
@@ -15,9 +15,13 @@
+
 .i-amphtml-accordion-content {
   margin: 0;
+  display: block !important;
```
I was wrong for this one and the one in the JSS file :( But we can't force `display: block` because `display: flex` is a perfectly valid value here. Everything else should work unchanged though. Let's just remove the forced `display: block`.

krdwan replied:
Updated in latest revision.  (removed `!important`). Also cleaned up the `css` file a bit

dvoytenko commented on extensions/amp-accordion/1.0/amp-accordion.css:
```diff
@@ -29,3 +33,25 @@
+  display: block !important;
+}
+
+amp-accordion > section:not([expanded]) > :last-child:not(.i-amphtml-animating),
```
Please bring these ones above the experimental version.

krdwan replied:
Updated in latest revision.

dvoytenko commented on extensions/amp-accordion/1.0/animations.js:
```diff
@@ -111,8 +112,7 @@ export function animateCollapse(content) {
     () => {
-      // Undo the `content.hidden = true` above.
-      content.hidden = true;
+      content.classList.remove('i-amphtml-animating');
```
Looks like `add` and `remove` for this class should go into the common `animate()` function below since it's always the same. But I could be wrong and something might not match up. But please try.

krdwan replied:
Updated in latest revision.

caroqliu commented on extensions/amp-accordion/1.0/accordion.js:
```diff
@@ -426,11 +489,13 @@ export function AccordionContent({
       aria-labelledby={headerId}
       role={role}
-      hidden={!expanded}
+      aria-hidden={String(!expanded)}
```
`aria-hidden` should not be necessary with `display: none` or `visibility: hidden`, so we should verify if it is needed for `content-visibility: hidden-matchable`. Played around with it and seems a bit buggy. cc @vmpstr as FYI

krdwan replied:
If it's buggy for `content-visibility: hidden-matchable` would it make sense to keep it to be safe?

dvoytenko replied:
Let's confirm with the spec folk. If intention is to make it implied with `hidden-matchable`, then, yes, I'm in favor of removing.

krdwan replied:
Removed for now.  Can add back in if we hear differently from @vmpstr.  Thanks!!

krdwan replied:
Confirmed with @caroqliu offline who has been in contact with @vmpstr that we do not need `aria-hidden` when `content-visibility: hidden-matchable` is applied to the element and that the observed buggy behavior is currently in progress to be fixed and should be in Canary shortly.

caroqliu commented on extensions/amp-accordion/1.0/accordion.type.js:
```diff
@@ -99,6 +100,7 @@ AccordionDef.ContentShimProps;
  *   toggleExpanded: (function(string)|undefined),
  *   animate: (boolean|undefined),
  *   prefix: (string),
+ *   experimentDisplayLocking: (boolean),
```
fyi for singular types, we don't need parens. It looks like they are used this way throughout this file so no need to change.

krdwan replied:
Thanks!

krdwan commented on extensions/amp-accordion/1.0/accordion.js:
```diff
@@ -299,22 +310,25 @@ export function AccordionSection({
+      } else {
+        setExpandedState((prev) => {
+          const newValue = opt_expand || !prev;
+          Promise.resolve().then(() => {
```
Sure, all of the code that fires events (basically any time that `onExpandStateChange` is called) is done inside a `Promise.resolve()` so that it can execute after the current render.  W/o explicitly forcing the event to run after the current thread, there could be a lot of timing issues.

For example - The accordion can support toggling all sections at once.  When a particular section is toggled it can send an event to say collpase (or expand) another section.  That other section is also going to be toggled by the original toggling all sections.  So which should resolve first?

So this way we explicitly say, all toggling happens.  THEN all events are triggered and further collapsing / expanding happens after the initial toggling.

dvoytenko replied:
To clarify, the async even call here is necessary because we're using `setState(function() {})` form of the set state. Unfortunately, that form (itself asynchronous) doesn't "like" even calls from it. But I think the promise here is ok for now. We'll likely come up with some nice `useAsyncCallback` hook eventually for this. This is rare, but we had a couple of cases where this happened already.

dvoytenko commented on extensions/amp-accordion/1.0/animations.js:
```diff
@@ -124,12 +109,14 @@ export function animateCollapse(content) {
     if (cleanup) {
       cleanup();
     }
+    element.classList.remove('i-amphtml-animating');
```
We also need to put this in `player.oncancel`.

krdwan replied:
As discussed offline, in this case onfinish and oncancel should both do the same and should both remove the class at the end! :)

dvoytenko reviewed (changes requested):
One important note on animation cleanup.

caroqliu commented on extensions/amp-accordion/1.0/accordion.js:
```diff
@@ -375,7 +397,7 @@ export function AccordionHeader({
       tabIndex={tabIndex}
       aria-controls={contentId}
-      onClick={expandHandler}
+      onClick={() => toggleHandler()}
```
nit: 
```suggestion
      onClick={toggleHandler}
```

krdwan replied:
The first argument `toggleHandler` accepts is the `opt_expand` (force expand or force collapse).  Wrapping the onClick callback in this function prevents the `click` event from being passed into `toggleHandler` as the first argument.

caroqliu commented on extensions/amp-accordion/1.0/amp-accordion.js:
```diff
@@ -242,7 +255,7 @@ const bindHeaderShimToElement = (element) => HeaderShim.bind(null, element);
 function ContentShimWithRef(
   sectionElement,
-  {hidden, id, role, 'aria-labelledby': ariaLabelledBy},
+  {id, role, 'aria-labelledby': ariaLabelledBy, 'aria-hidden': ariaHidden},
```
Do we need to remove `aria-hidden` inclusions from this file?

krdwan replied:
Thanks, yep, I did another pass which removes from the amp-side, thanks!

caroqliu commented on extensions/amp-accordion/1.0/test/test-accordion.js:
```diff
@@ -73,7 +73,7 @@ describes.sandboxed('Accordion preact component', {}, (env) => {
       expect(content.localName).to.equal('div');
       expect(content.innerHTML).to.equal('content1');
-      expect(content.hidden).to.be.false;
+      expect(content.getAttribute('aria-hidden')).to.equal('false');
```
Looks like these lines have to be removed as well

krdwan replied:
Yep latest commit has updates to tests which includes this line.

caroqliu commented on extensions/amp-accordion/1.0/test/test-amp-accordion.js:
```diff
@@ -516,6 +517,94 @@ describes.realWin(
+        const section1 = element.children[0];
+        const content1 = section1.lastElementChild;
+
+        expect(win.AMP.isExperimentOn('amp-accordion-display-locking')).to.be
```
Do we need this expect statement if we know we turn on the experiment in the `beforeEach` block?

krdwan replied:
I don't think it hurts?

caroqliu replied:
I don't think it helps. IMO we should remove so the tests can concisely add confidence. But it's up to you.

caroqliu commented on extensions/amp-accordion/1.0/test/test-amp-accordion.js:
```diff
@@ -516,6 +517,94 @@ describes.realWin(
+      });
+
+      afterEach(() => {
+        win.CSS.supports = defaultCssSupports;
```
please also toggleExperiment off in this block

krdwan replied:
Added in the latest.

caroqliu reviewed (commented):
Just a few more comments!
